<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Jean Morrison" />

<meta name="date" content="2020-05-14" />

<title>Summary of Relevant Literature</title>

<script src="site_libs/header-attrs-2.3/header-attrs.js"></script>
<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/master/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>



<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>




<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 51px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h2 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h3 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h4 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h5 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h6 {
  padding-top: 56px;
  margin-top: -56px;
}
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #ffffff;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  background: white;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row-fluid">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">gwas_factors</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://gitlab.com/jean_morrison/gwas-summay-statistic-factorization">
    <span class="fa fa-gitlab"></span>
     
    Source code
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div class="fluid-row" id="header">



<h1 class="title toc-ignore">Summary of Relevant Literature</h1>
<h4 class="author">Jean Morrison</h4>
<h4 class="date">2020-05-14</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span> workflowr <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> </a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2020-10-13
</p>
<p>
<strong>Checks:</strong> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 2 <span class="glyphicon glyphicon-exclamation-sign text-danger" aria-hidden="true"></span> 0
</p>
<p>
<strong>Knit directory:</strong> <code>sumstatFactors/</code> <span class="glyphicon glyphicon-question-sign" aria-hidden="true" title="This is the local directory in which the code in this file was executed."> </span>
</p>
<p>
This reproducible <a href="http://rmarkdown.rstudio.com">R Markdown</a> analysis was created with <a
  href="https://github.com/jdblischak/workflowr">workflowr</a> (version 1.6.2). The <em>Checks</em> tab describes the reproducibility checks that were applied when the results were created. The <em>Past versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguptodate"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>R Markdown file:</strong> up-to-date </a>
</p>
</div>
<div id="strongRMarkdownfilestronguptodate" class="panel-collapse collapse">
<div class="panel-body">
<p>Great! Since the R Markdown file has been committed to the Git repository, you know the exact version of the code that produced these results.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgitlabcomjeanmorrisongwassummaystatisticfactorizationtree80f2d339c7f4ea2ef17c82a83c17bdf8c774ca44targetblank80f2d33a"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Repository version:</strong> <a href="https://gitlab.com/jean_morrison/gwas-summay-statistic-factorization/tree/80f2d339c7f4ea2ef17c82a83c17bdf8c774ca44" target="_blank">80f2d33</a> </a>
</p>
</div>
<div id="strongRepositoryversionstrongahrefhttpsgitlabcomjeanmorrisongwassummaystatisticfactorizationtree80f2d339c7f4ea2ef17c82a83c17bdf8c774ca44targetblank80f2d33a" class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development and connecting the code version to the results is critical for reproducibility.
</p>
<p>
The results in this page were generated with repository version <a href="https://gitlab.com/jean_morrison/gwas-summay-statistic-factorization/tree/80f2d339c7f4ea2ef17c82a83c17bdf8c774ca44" target="_blank">80f2d33</a>. See the <em>Past versions</em> tab to see a history of the changes made to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for the analysis have been committed to Git prior to generating the results (you can use <code>wflow_publish</code> or <code>wflow_git_commit</code>). workflowr only checks the R Markdown file, but you know if there are other scripts or data files that it depends on. Below is the status of the Git repository when the results were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .Rhistory
    Ignored:    .Rproj.user/
    Ignored:    analysis/fixed_factors2_cache/
    Ignored:    analysis/fixed_factors_cache/
    Ignored:    analysis/metabo_flash_experiments_cache/
    Ignored:    analysis/pathway_factors_cache/
    Ignored:    analysis/simulations_cache/
    Ignored:    analysis/z_scores_cache/
    Ignored:    analysis_data/2020-06-10_ai/
    Ignored:    analysis_data/2020-10-02_metabolites/
    Ignored:    analysis_data/simple_patterns/
    Ignored:    analysis_data/sims_medium_k/

Untracked files:
    Untracked:  analysis/sample_corr2.Rmd
    Untracked:  buttons.css
    Untracked:  bw_factors2.png
    Untracked:  code/hide_output.js
    Untracked:  k99_analysis.R
    Untracked:  k99_fit.RDS
    Untracked:  matrix_20.png

Unstaged changes:
    Modified:   analysis/k_greater_than_n.Rmd
    Modified:   analysis/sumstat_factors.bib

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not included in this status report because it is ok for generated content to have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">

<p>
These are the previous versions of the repository in which changes were made to the R Markdown (<code>analysis/literature.Rmd</code>) and HTML (<code>public/literature.html</code>) files. If you’ve configured a remote Git repository (see <code>?wflow_git_remote</code>), click on the hyperlinks in the table below to view the files as they were in that past version.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://gitlab.com/jean_morrison/gwas-summay-statistic-factorization/blob/f6f6e9cee677e31a339b7ef089438d85e49b69ca/analysis/literature.Rmd" target="_blank">f6f6e9c</a>
</td>
<td>
Jean Morrison
</td>
<td>
2020-10-12
</td>
<td>
Add kettunen et al
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://glcdn.githack.com/jean_morrison/gwas-summay-statistic-factorization/raw/f3643cebf98fa0fb754c5751a21447db21367701/public/literature.html" target="_blank">f3643ce</a>
</td>
<td>
Jean Morrison
</td>
<td>
2020-09-25
</td>
<td>
Host with GitLab.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://gitlab.com/jean_morrison/gwas-summay-statistic-factorization/blob/42cdd843b43d86267ef86a2ee4616aa322497c67/analysis/literature.Rmd" target="_blank">42cdd84</a>
</td>
<td>
Jean Morrison
</td>
<td>
2020-09-16
</td>
<td>
track lit review
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<div id="background" class="section level2">
<h2>Background</h2>
<ul>
<li><p>FLASH</p></li>
<li><p>penalized SVD</p></li>
<li><p><span class="citation">Watanabe et al. (2019)</span> A global overview of pleiotropy and genetic architecture in complex traits</p></li>
</ul>
</div>
<div id="related-work" class="section level2">
<h2>Related work</h2>
<ul>
<li><p><span class="citation">Urbut et al. (2018)</span> MASH</p></li>
<li><p><span class="citation">Tanigawa et al. (2019)</span> Components of genetic associations across 2,138 phenotypes in the UK Biobank highlight adipocyte biology</p></li>
</ul>
<p>Truncated SVD via DeGAs. Separate analysis on different variant sets (1) all vairants, (2) coding vairants (3) protein truncating variants. Retained z-scores for variant-trait pairs with <span class="math inline">\(p &lt; 0.001\)</span> and standard error below a threshold. I believe other z-scores were set to zero. I don’t think there is any correction for sample overlap.</p>
<ul>
<li><span class="citation">Argelaguet et al. (2018)</span> Multi-Omics Factor Analysis (MOFA)</li>
</ul>
<p>Intended for integrating different (high dimensioanal) data modalities (e.g. expression, methylation etc), uses individual level data. Loadings can be sparse. Based on group factor analysis (a long list of references! Virtanen et al 2012, Khan et al 2014, Klami et al 2015, Bunte et al 2016, Zhao et al 2016, Leppaaho and Kaski 2017). “(i) fast inference based on a varia- tional approximation, (ii) inference of sparse solutions facilitating interpretation, (iii) efficient handling of missing values and (iv) flex- ible combination of different likelihood models for each data modality, which enables integrating diverse data types such as binary-, count- and continuous-valued data.”</p>
<ul>
<li><p><span class="citation">Burren et al. (2020)</span> Characterisation of the genetic architecture of immune mediated disease through informed dimension reduction. Uses PCA</p></li>
<li><p><span class="citation">Grotzinger et al. (2019)</span> Genomic SEM Provides Insights into the Multivariate Genetic Architecture of Complex Traits</p></li>
<li><p><span class="citation">Runcie and Mukherjee (2013)</span> Dissecting high-dimensional phenotypes with Bayesian sparse factor analysis of genetic covariance matrices</p></li>
<li><p><span class="citation">Chasman et al. (2020)</span> Pleiotropy based decomposition of genomic risk scores.</p></li>
<li><p><span class="citation">Gao, Brown, and Engelhardt (2013)</span> A latent factor model with a mixture of sparse and dense factors to model gene expression data with confounding effects</p></li>
</ul>
<p>Based on SVD</p>
</div>
<div id="data-applications" class="section level2">
<h2>Data Applications</h2>
<ul>
<li><p><span class="citation">Kettunen et al. (2016)</span> Metabolites</p></li>
<li><p><span class="citation">Astle et al. (2016)</span> Blood cells data set</p></li>
<li><p><span class="citation">Khramtsova, Davis, and Stranger (2019)</span> Review article about sex differences in common diseases and common disease genetics.</p></li>
</ul>
</div>
<div id="references" class="section level2 unnumbered">
<h2 class="unnumbered">References</h2>
<div id="refs" class="references hanging-indent">
<div id="ref-Argelaguet2018">
<p>Argelaguet, Ricard, Britta Velten, Damien Arnol, Sascha Dietrich, Thorsten Zenz, John C Marioni, Florian Buettner, Wolfgang Huber, and Oliver Stegle. 2018. “Multi‐Omics Factor Analysis—a framework for unsupervised integration of multi‐omics data sets.” <em>Molecular Systems Biology</em> 14: e8124. <a href="https://doi.org/10.15252/msb.20178124">https://doi.org/10.15252/msb.20178124</a>.</p>
</div>
<div id="ref-Astle2016">
<p>Astle, William J., Heather Elding, Tao Jiang, Dave Allen, Dace Ruklisa, Alice L. Mann, Daniel Mead, et al. 2016. “The Allelic Landscape of Human Blood Cell Trait Variation and Links to Common Complex Disease.” <em>Cell</em> 167 (5): 1415–1429.e19. <a href="https://doi.org/10.1016/j.cell.2016.10.042">https://doi.org/10.1016/j.cell.2016.10.042</a>.</p>
</div>
<div id="ref-Burren2020">
<p>Burren, Oliver S., Guillermo Reales, Limy Wong, John Bowes, James C Lee, Anne Bartion, Paul A Lyons, et al. 2020. “Characterisation of the genetic architecture of immune mediated disease through informed dimension reduction.” <em>bioRxiv</em>, 1–31.</p>
</div>
<div id="ref-Chasman2020">
<p>Chasman, Daniel I., Franco Giulianini, Olga V. Demler, and Miriam S. Udler. 2020. “Pleiotropy-Based Decomposition of Genetic Risk Scores: Association and Interaction Analysis for Type 2 Diabetes and CAD.” <em>American Journal of Human Genetics</em> 106 (5): 646–58. <a href="https://doi.org/10.1016/j.ajhg.2020.03.011">https://doi.org/10.1016/j.ajhg.2020.03.011</a>.</p>
</div>
<div id="ref-Gao2013">
<p>Gao, Chuan, Christopher D Brown, and Barbara E Engelhardt. 2013. “A latent factor model with a mixture of sparse and dense factors to model gene expression data with confounding effects.” <em>arXiv</em>, 1–28. <a href="http://arxiv.org/abs/1310.4792">http://arxiv.org/abs/1310.4792</a>.</p>
</div>
<div id="ref-Grotzinger2019">
<p>Grotzinger, Andrew D, Mijke Rhemtulla, Ronald de Vlaming, Stuart J Ritchie, T Mallard, W David Hill, Hill F Ip, et al. 2019. “Genomic SEM Provides Insights into the Multivariate Genetic Architecture of Complex Traits.” <em>Nature Human Behaviour</em> 3 (5): 513–25. <a href="https://doi.org/10.1038/s41562-019-0566-x.Genomic">https://doi.org/10.1038/s41562-019-0566-x.Genomic</a>.</p>
</div>
<div id="ref-Kettunen2016">
<p>Kettunen, Johannes, Ayse Demirkan, Peter Würtz, Harmen H. M. Draisma, Toomas Haller, Rajesh Rawal, Anika Vaarhorst, et al. 2016. “Genome-wide study for circulating metabolites identifies 62 loci and reveals novel systemic effects of LPA.” <em>Nature Communications</em> 7: 1–9. <a href="https://doi.org/10.1038/ncomms11122">https://doi.org/10.1038/ncomms11122</a>.</p>
</div>
<div id="ref-Khramtsova2019">
<p>Khramtsova, Ekaterina A., Lea K. Davis, and Barbara E. Stranger. 2019. “The role of sex in the genomics of human complex traits.” <em>Nature Reviews Genetics</em> 20 (3): 173–90. <a href="https://doi.org/10.1038/s41576-018-0083-1">https://doi.org/10.1038/s41576-018-0083-1</a>.</p>
</div>
<div id="ref-Runcie2013">
<p>Runcie, Daniel E., and Sayan Mukherjee. 2013. “Dissecting high-dimensional phenotypes with Bayesian sparse factor analysis of genetic covariance matrices.” <em>Genetics</em> 194 (3): 753–67. <a href="https://doi.org/10.1534/genetics.113.151217">https://doi.org/10.1534/genetics.113.151217</a>.</p>
</div>
<div id="ref-Tanigawa2019">
<p>Tanigawa, Yosuke, Jiehan Li, Johanne M. Justesen, Heiko Horn, Matthew Aguirre, Christopher DeBoever, Chris Chang, et al. 2019. “Components of genetic associations across 2,138 phenotypes in the UK Biobank highlight adipocyte biology.” <em>Nature Communications</em> 10 (1). <a href="https://doi.org/10.1038/s41467-019-11953-9">https://doi.org/10.1038/s41467-019-11953-9</a>.</p>
</div>
<div id="ref-Urbut2018">
<p>Urbut, Sarah M., Gao Wang, Peter Carbonetto, and Matthew Stephens. 2018. “Flexible statistical methods for estimating and testing effects in genomic studies with multiple conditions.” <em>Nature Genetics</em> 51: 187–95. <a href="https://doi.org/10.1038/s41588-018-0268-8">https://doi.org/10.1038/s41588-018-0268-8</a>.</p>
</div>
<div id="ref-Watanabe2019">
<p>Watanabe, Kyoko, Sven Stringer, Oleksandr Frei, Maša Umićević Mirkov, Christiaan de Leeuw, Tinca J. C. Polderman, Sophie van der Sluis, Ole A. Andreassen, Benjamin M. Neale, and Danielle Posthuma. 2019. “A global overview of pleiotropy and genetic architecture in complex traits.” <em>Nature Genetics</em> 51 (9): 1339–48. <a href="https://doi.org/10.1038/s41588-019-0481-0">https://doi.org/10.1038/s41588-019-0481-0</a>.</p>
</div>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
http://docs.mathjax.org/en/latest/configuration.html.  This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>




</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open')
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
