<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="GFA">
<title>Running GFA with Simulated Data â€¢ GFA</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.11/clipboard.min.js" integrity="sha512-7O5pXpc0oCRrxk8RUfDYFgn0nO1t+jLuIOQdOMRp4APB7uZ4vSjspzp5y6YDtDs4VzUSTbWzBFZ/LKJhnyFOKw==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Running GFA with Simulated Data">
<meta property="og:description" content="GFA">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light" data-bs-theme="light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">GFA</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0.0.0445</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/gfa_with_simulated_data.html">Running GFA with Simulated Data</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav"></ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Running GFA with Simulated Data</h1>
            
      
      
      <div class="d-none name"><code>gfa_with_simulated_data.Rmd</code></div>
    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://jean997.github.io/GFA/">GFA</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/jean997/GWASBrewer" class="external-link">GWASBrewer</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'dplyr'</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:stats':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     filter, lag</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:base':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     intersect, setdiff, setequal, union</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://sjmgarnier.github.io/viridis/" class="external-link">viridis</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; Loading required package: viridisLite</span></span></code></pre></div>
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>In this vignette we will simulate two toy data sets and analyze them
with GFA. The first toy data set is very simple and is the example shown
in Figure 1 of the paper. The second toy data set is more complex. It
includes more traits, a more complicated factor structure, and data are
simulated with LD. For the second data set, we will compare results with
and without accounting for correlation due to overlapping samples.</p>
</div>
<div class="section level2">
<h2 id="simple-example">Simple Example<a class="anchor" aria-label="anchor" href="#simple-example"></a>
</h2>
<div class="section level3">
<h3 id="simulate-simple-data">Simulate simple data<a class="anchor" aria-label="anchor" href="#simulate-simple-data"></a>
</h3>
<p>We will start by constructing and plotting the true factor structure.
GFA includes a convenience function <code>plot_factors</code> for
plotting matrices as heatmaps.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">100</span><span class="op">)</span></span>
<span><span class="co"># True factor structure</span></span>
<span><span class="va">myF</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="op">-</span><span class="fl">1</span>, <span class="fl">1</span>, <span class="fl">1</span>, <span class="fl">2</span>, <span class="fl">1</span><span class="op">)</span>, nrow <span class="op">=</span> <span class="fl">3</span>, byrow <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="va">myF</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/apply.html" class="external-link">apply</a></span><span class="op">(</span><span class="va">myF</span>, <span class="fl">2</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">{</span><span class="va">x</span><span class="op">/</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">sqrt</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">x</span><span class="op">^</span><span class="fl">2</span><span class="op">)</span><span class="op">)</span><span class="op">}</span><span class="op">)</span></span>
<span></span>
<span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plot_factors.html">plot_factors</a></span><span class="op">(</span><span class="va">myF</span>, row_names <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"T3"</span>, <span class="st">"T2"</span>, <span class="st">"T1"</span><span class="op">)</span>, col_names <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"F1"</span>, <span class="st">"F2"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>     <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_gradient.html" class="external-link">scale_fill_gradient2</a></span><span class="op">(</span>low <span class="op">=</span> <span class="fu"><a href="https://sjmgarnier.github.io/viridisLite/reference/viridis.html" class="external-link">viridis</a></span><span class="op">(</span><span class="fl">3</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, high <span class="op">=</span> <span class="fu"><a href="https://sjmgarnier.github.io/viridisLite/reference/viridis.html" class="external-link">viridis</a></span><span class="op">(</span><span class="fl">3</span><span class="op">)</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span>, name <span class="op">=</span> <span class="st">"Normalized\nEffect"</span><span class="op">)</span> <span class="op">+</span></span>
<span>     <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">xlab</a></span><span class="op">(</span><span class="st">"Mediator"</span><span class="op">)</span> <span class="op">+</span></span>
<span>     <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"True Mediator Network"</span><span class="op">)</span> <span class="op">+</span></span>
<span>     <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_discrete.html" class="external-link">scale_x_discrete</a></span><span class="op">(</span>position <span class="op">=</span> <span class="st">"top"</span><span class="op">)</span><span class="op">+</span></span>
<span>     <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>axis.text.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">0</span><span class="op">)</span>,</span>
<span>           panel.background <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_rect</a></span><span class="op">(</span>fill <span class="op">=</span> <span class="st">"white"</span><span class="op">)</span>,</span>
<span>           axis.ticks <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; Scale for <span style="color: #00BB00;">fill</span> is already present.</span></span>
<span><span class="co">#&gt; Adding another scale for <span style="color: #00BB00;">fill</span>, which will replace the existing scale.</span></span>
<span><span class="va">p</span></span></code></pre></div>
<p><img src="gfa_with_simulated_data_files/figure-html/unnamed-chunk-2-1.png" width="700"></p>
<p>Next we simulate some data using the <code>sim_lf</code> function in
<code>GWASBrewer</code> which simulates from the GFA model,</p>
<p><span class="math display">\[
\hat{B} = LF^\top + \Theta + E
\]</span> <span class="math inline">\(F\)</span> is the factor
structure, <span class="math inline">\(L\)</span> is the randomly
generated matrix of variant-factor effects, <span class="math inline">\(\Theta\)</span> is an additional matrix of direct
trait effects, and <span class="math inline">\(E\)</span> is measurement
error.</p>
<p>Our first example is a toy example designed to have a very strong
signal that can be seen visually, but we donâ€™t consider these parameters
realistic. We use three traits that each have a heritability of 0.5,
GWAS sample size of <span class="math inline">\(50,000\)</span>, and 80%
of trait heritability explained by the two factors. We generate 10k
independent SNPs (no LD). <span class="math inline">\(L\)</span> and
<span class="math inline">\(\Theta\)</span> both have 10% non-zero
elements. There is no overlap between the GWAS samples.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dat</span> <span class="op">&lt;-</span> <span class="fu">GWASBrewer</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/GWASBrewer/man/sim_lf.html" class="external-link">sim_lf</a></span><span class="op">(</span>F_mat <span class="op">=</span> <span class="va">myF</span>, </span>
<span>                          N <span class="op">=</span> <span class="fl">50000</span>, </span>
<span>                          J <span class="op">=</span> <span class="fl">10000</span>, </span>
<span>                          h2_trait <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">0.5</span>, <span class="fl">3</span><span class="op">)</span>,</span>
<span>                          omega <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">0.8</span>, <span class="fl">3</span><span class="op">)</span>, </span>
<span>                          pi_L <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">0.1</span>,<span class="fl">2</span><span class="op">)</span>, </span>
<span>                          pi_theta <span class="op">=</span> <span class="fl">0.1</span>, </span>
<span>                          est_s <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co">#&gt; SNP effects provided for 10000 SNPs and 3 traits.</span></span></code></pre></div>
<p>For illustrative purposes, we plot some of the simulated effect
estimates <span class="math inline">\(\beta_hat\)</span>, selecting SNPS
acting through a range of possible pathways with large effect sizes so
that the pattern is clear visually.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Null SNPS which do not effect factors or have additional effects in theta</span></span>
<span><span class="va">ix0</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">rowSums</a></span><span class="op">(</span><span class="va">dat</span><span class="op">$</span><span class="va">L_mat_joint</span> <span class="op">==</span> <span class="fl">0</span><span class="op">)</span> <span class="op">==</span> <span class="fl">2</span><span class="op">)</span> <span class="op">&amp;</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">rowSums</a></span><span class="op">(</span><span class="va">dat</span><span class="op">$</span><span class="va">theta_joint</span> <span class="op">==</span> <span class="fl">0</span><span class="op">)</span> <span class="op">==</span> <span class="fl">3</span><span class="op">)</span><span class="op">)</span>, size <span class="op">=</span> <span class="fl">8</span><span class="op">)</span></span>
<span><span class="co"># SNPS which only effect trait 1 directly and not through factors</span></span>
<span><span class="va">ix1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">rowSums</a></span><span class="op">(</span><span class="va">dat</span><span class="op">$</span><span class="va">L_mat_joint</span> <span class="op">==</span> <span class="fl">0</span><span class="op">)</span> <span class="op">==</span> <span class="fl">2</span><span class="op">)</span> <span class="op">&amp;</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">rowSums</a></span><span class="op">(</span><span class="va">dat</span><span class="op">$</span><span class="va">theta_joint</span> <span class="op">==</span> <span class="fl">0</span><span class="op">)</span> <span class="op">==</span> <span class="fl">2</span><span class="op">)</span> <span class="op">&amp;</span> <span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">abs</a></span><span class="op">(</span><span class="va">dat</span><span class="op">$</span><span class="va">theta_joint</span><span class="op">[</span>,<span class="fl">1</span><span class="op">]</span><span class="op">)</span><span class="op">&gt;</span> <span class="fl">0.02</span><span class="op">)</span>, size <span class="op">=</span> <span class="fl">4</span><span class="op">)</span></span>
<span><span class="co"># SNPS which only effect trait 2 directly and not through factors</span></span>
<span><span class="va">ix2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">rowSums</a></span><span class="op">(</span><span class="va">dat</span><span class="op">$</span><span class="va">L_mat_joint</span> <span class="op">==</span> <span class="fl">0</span><span class="op">)</span> <span class="op">==</span> <span class="fl">2</span><span class="op">)</span> <span class="op">&amp;</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">rowSums</a></span><span class="op">(</span><span class="va">dat</span><span class="op">$</span><span class="va">theta_joint</span> <span class="op">==</span> <span class="fl">0</span><span class="op">)</span> <span class="op">==</span> <span class="fl">2</span><span class="op">)</span> <span class="op">&amp;</span> <span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">abs</a></span><span class="op">(</span><span class="va">dat</span><span class="op">$</span><span class="va">theta_joint</span><span class="op">[</span>,<span class="fl">2</span><span class="op">]</span><span class="op">)</span><span class="op">&gt;</span> <span class="fl">0.02</span><span class="op">)</span>, size <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span>
<span><span class="co"># SNPS which only effect trait 3 directly and not through factors</span></span>
<span><span class="va">ix3</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">rowSums</a></span><span class="op">(</span><span class="va">dat</span><span class="op">$</span><span class="va">L_mat_joint</span> <span class="op">==</span> <span class="fl">0</span><span class="op">)</span> <span class="op">==</span> <span class="fl">2</span><span class="op">)</span> <span class="op">&amp;</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">rowSums</a></span><span class="op">(</span><span class="va">dat</span><span class="op">$</span><span class="va">theta_joint</span> <span class="op">==</span> <span class="fl">0</span><span class="op">)</span> <span class="op">==</span> <span class="fl">2</span><span class="op">)</span> <span class="op">&amp;</span> <span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">abs</a></span><span class="op">(</span><span class="va">dat</span><span class="op">$</span><span class="va">theta_joint</span><span class="op">[</span>,<span class="fl">3</span><span class="op">]</span><span class="op">)</span><span class="op">&gt;</span> <span class="fl">0.02</span><span class="op">)</span>, size <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span>
<span><span class="co"># SNPS which only effect effect factor 1 </span></span>
<span><span class="va">ix4</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">rowSums</a></span><span class="op">(</span><span class="va">dat</span><span class="op">$</span><span class="va">L_mat_joint</span> <span class="op">==</span> <span class="fl">0</span><span class="op">)</span> <span class="op">==</span> <span class="fl">1</span><span class="op">)</span> <span class="op">&amp;</span> <span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">abs</a></span><span class="op">(</span><span class="va">dat</span><span class="op">$</span><span class="va">L_mat_joint</span><span class="op">[</span>,<span class="fl">1</span><span class="op">]</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">0.07</span> <span class="op">&amp;</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">rowSums</a></span><span class="op">(</span><span class="va">dat</span><span class="op">$</span><span class="va">theta_joint</span> <span class="op">==</span> <span class="fl">0</span><span class="op">)</span> <span class="op">==</span> <span class="fl">3</span><span class="op">)</span><span class="op">)</span>, size <span class="op">=</span> <span class="fl">6</span><span class="op">)</span></span>
<span><span class="co"># SNPS which only effect factor 2 </span></span>
<span><span class="va">ix5</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">rowSums</a></span><span class="op">(</span><span class="va">dat</span><span class="op">$</span><span class="va">L_mat_joint</span> <span class="op">==</span> <span class="fl">0</span><span class="op">)</span> <span class="op">==</span> <span class="fl">1</span><span class="op">)</span> <span class="op">&amp;</span> <span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">abs</a></span><span class="op">(</span><span class="va">dat</span><span class="op">$</span><span class="va">L_mat_joint</span><span class="op">[</span>,<span class="fl">2</span><span class="op">]</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">0.07</span> <span class="op">&amp;</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">rowSums</a></span><span class="op">(</span><span class="va">dat</span><span class="op">$</span><span class="va">theta_joint</span> <span class="op">==</span> <span class="fl">0</span><span class="op">)</span> <span class="op">==</span> <span class="fl">3</span><span class="op">)</span><span class="op">)</span>, size <span class="op">=</span> <span class="fl">4</span><span class="op">)</span></span>
<span></span>
<span><span class="va">X</span> <span class="op">&lt;-</span> <span class="va">dat</span><span class="op">$</span><span class="va">beta_hat</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">ix0</span>, <span class="va">ix1</span>, <span class="va">ix2</span>, <span class="va">ix3</span>, <span class="va">ix4</span>, <span class="va">ix5</span><span class="op">)</span>,<span class="op">]</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">3</span>,<span class="fl">2</span>,<span class="fl">1</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="va">se</span> <span class="op">&lt;-</span> <span class="va">dat</span><span class="op">$</span><span class="va">s_estimate</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">ix0</span>, <span class="va">ix1</span>, <span class="va">ix2</span>, <span class="va">ix3</span>, <span class="va">ix4</span>, <span class="va">ix5</span><span class="op">)</span>,<span class="op">]</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">3</span>,<span class="fl">2</span>,<span class="fl">1</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="fu"><a href="../reference/plot_factors.html">plot_factors</a></span><span class="op">(</span><span class="va">X</span>, col_names <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"T1"</span>, <span class="st">"T2"</span>, <span class="st">"T3"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_gradient.html" class="external-link">scale_fill_gradient2</a></span><span class="op">(</span>low <span class="op">=</span> <span class="fu"><a href="https://sjmgarnier.github.io/viridisLite/reference/viridis.html" class="external-link">viridis</a></span><span class="op">(</span><span class="fl">3</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, high <span class="op">=</span> <span class="fu"><a href="https://sjmgarnier.github.io/viridisLite/reference/viridis.html" class="external-link">viridis</a></span><span class="op">(</span><span class="fl">3</span><span class="op">)</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span>, name <span class="op">=</span> <span class="st">"Beta hat"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">xlab</a></span><span class="op">(</span><span class="st">"Trait"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ylab</a></span><span class="op">(</span><span class="st">"Variant"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"Variant-Trait Associations"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_discrete.html" class="external-link">scale_x_discrete</a></span><span class="op">(</span>position <span class="op">=</span> <span class="st">"top"</span><span class="op">)</span><span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>axis.text.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">0</span><span class="op">)</span>,</span>
<span>           panel.background <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_rect</a></span><span class="op">(</span>fill <span class="op">=</span> <span class="st">"white"</span><span class="op">)</span>,</span>
<span>           axis.ticks <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; Scale for <span style="color: #00BB00;">fill</span> is already present.</span></span>
<span><span class="co">#&gt; Adding another scale for <span style="color: #00BB00;">fill</span>, which will replace the existing scale.</span></span></code></pre></div>
<p><img src="gfa_with_simulated_data_files/figure-html/unnamed-chunk-4-1.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="fit-gfa-to-simple-data">Fit GFA to simple data<a class="anchor" aria-label="anchor" href="#fit-gfa-to-simple-data"></a>
</h3>
<p>In a typical GFA analysis of real data, the next step is to estimate
the nuisance correlation matrix. In our toy example, there is no sample
overlap, so we will skip this step, but we will include it in the next
example.</p>
<p>To fit GFA to our toy data, we use the <code>gfa_fit</code> function.
<code>gfa_fit</code> can take two types of input, either z-scores and
sample size or effect estimates and standard errors. If you know the
sample size for each GWAS, it is better to use z-scores and sample size.
In either case, GFA will be fit using z-scores and the fitted factors
will be re-scaled so that they are on the standardized trait scale. If
z-scores are provided and no sample size is given, factors can still be
estimated, but will be left on the z-score scale which depends on sample
size and may not be interpretable. We will discuss scaling in greater
detail later in this vignette.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">Z_hat</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/with.html" class="external-link">with</a></span><span class="op">(</span><span class="va">dat</span>, <span class="va">beta_hat</span><span class="op">/</span><span class="va">s_estimate</span><span class="op">)</span></span>
<span><span class="va">fit_toy</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/gfa_fit.html">gfa_fit</a></span><span class="op">(</span>Z_hat <span class="op">=</span> <span class="va">Z_hat</span>, </span>
<span>                   N <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">50000</span>, <span class="fl">3</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; R is not supplied, fitting assuming full independence</span></span>
<span><span class="co">#&gt; Adding factor 1 to flash object...</span></span>
<span><span class="co">#&gt; Adding factor 2 to flash object...</span></span>
<span><span class="co">#&gt; Adding factor 3 to flash object...</span></span>
<span><span class="co">#&gt; Adding factor 4 to flash object...</span></span>
<span><span class="co">#&gt; Factor doesn't significantly increase objective and won't be added.</span></span>
<span><span class="co">#&gt; Wrapping up...</span></span>
<span><span class="co">#&gt; Done.</span></span>
<span><span class="co">#&gt; Backfitting 3 factors (tolerance: 4.47e-04)...</span></span>
<span><span class="co">#&gt;   Difference between iterations is within 1.0e+02...</span></span>
<span><span class="co">#&gt;   Difference between iterations is within 1.0e+01...</span></span>
<span><span class="co">#&gt;   Difference between iterations is within 1.0e+00...</span></span>
<span><span class="co">#&gt;   Difference between iterations is within 1.0e-01...</span></span>
<span><span class="co">#&gt;   Difference between iterations is within 1.0e-02...</span></span>
<span><span class="co">#&gt;   Difference between iterations is within 1.0e-03...</span></span>
<span><span class="co">#&gt; Wrapping up...</span></span>
<span><span class="co">#&gt; Done.</span></span>
<span><span class="co">#&gt; Nullchecking 3 factors...</span></span>
<span><span class="co">#&gt; Done.</span></span>
<span><span class="co">#&gt; begin while</span></span>
<span><span class="co">#&gt; Nullchecking 3 factors...</span></span>
<span><span class="co">#&gt; Done.</span></span></code></pre></div>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">fit_toy</span><span class="op">)</span></span>
<span><span class="co">#&gt;  [1] "fit"          "method"       "L_hat"        "F_hat"        "F_hat_single"</span></span>
<span><span class="co">#&gt;  [6] "F_hat_multi"  "scale"        "gfa_pve"      "mode"         "params"</span></span></code></pre></div>
<p>The resulting fit object contains the following elements:</p>
<ul>
<li>
<code>F_hat</code>: estimated factor structure on standardized trait
scale</li>
<li>
<code>L_hat</code>: Estimated variant-factor effects</li>
<li>
<code>F_hat_single</code>: Subset of columns of <code>F_hat</code>
corresponding to single-trait factors.</li>
<li>
<code>F_hat_multi</code>: Columns of <code>F_hat</code>
corresponding to multi-trait factors.</li>
<li>
<code>F_hat_est</code>: estimated factors on the estimated (z-score)
scale (usually not needed).</li>
<li>
<code>gfa_pve</code>: An object containing <code>genet_var</code>
and <code>pve</code>. <code>genet_var</code> is the trait heritability
explained by the fitted model. <code>pve</code> is the proportion of
heritability for each trait (rows) explained by each factor
(columns).</li>
<li>
<code>scale</code>, <code>method</code>, <code>params</code> and
<code>fit</code> are for internal use.</li>
</ul>
<p>We can plot the estimated factors using the <code>plot_factors</code>
function:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fit_toy</span><span class="op">$</span><span class="va">F_hat</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="../reference/plot_factors.html">plot_factors</a></span><span class="op">(</span><span class="va">.</span>, col_names <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"EF1"</span>, <span class="st">"EF2"</span>, <span class="st">"EF3"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_gradient.html" class="external-link">scale_fill_gradient2</a></span><span class="op">(</span>low <span class="op">=</span> <span class="fu"><a href="https://sjmgarnier.github.io/viridisLite/reference/viridis.html" class="external-link">viridis</a></span><span class="op">(</span><span class="fl">3</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, high <span class="op">=</span> <span class="fu"><a href="https://sjmgarnier.github.io/viridisLite/reference/viridis.html" class="external-link">viridis</a></span><span class="op">(</span><span class="fl">3</span><span class="op">)</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span>, name <span class="op">=</span> <span class="st">"Normalized\nEffect"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">xlab</a></span><span class="op">(</span><span class="st">"Mediator"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ylab</a></span><span class="op">(</span><span class="st">"Trait"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"Estimated Mediator Network (GFA)"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_discrete.html" class="external-link">scale_x_discrete</a></span><span class="op">(</span>position <span class="op">=</span> <span class="st">"top"</span><span class="op">)</span><span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>axis.text.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">0</span><span class="op">)</span>,</span>
<span>           panel.background <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_rect</a></span><span class="op">(</span>fill <span class="op">=</span> <span class="st">"white"</span><span class="op">)</span>,</span>
<span>           axis.ticks <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; Scale for <span style="color: #00BB00;">fill</span> is already present.</span></span>
<span><span class="co">#&gt; Adding another scale for <span style="color: #00BB00;">fill</span>, which will replace the existing scale.</span></span></code></pre></div>
<p><img src="gfa_with_simulated_data_files/figure-html/unnamed-chunk-7-1.png" width="700"></p>
<p>The factors capture the truth quite well. Note that our estimated
factors are in a different order than the original. It is important to
keep in mind that the ordering of the factors is arbitrary. The sign of
each column is also arbitrary. We also have a third factor that effects
(almost) only trait 2. This captures direct effects on trait 2 that are
not mediated by the factors. For convenience, GFA will try to identify
these single trait factors and separate them, dividing
<code>F_hat</code> into <code>F_single</code> and <code>F_multi</code>
as described above. GFA is not restricted to produce a number of factors
equal to the number of traits. This occurs in this example by
coincidence.</p>
<p>For comparison we can try svd applied to z-scores for the most
significant variants which decomposes <span class="math inline">\(\hat{Z}\)</span> as <span class="math inline">\(UDV^\top\)</span> . Here the equivalent estimate
to our factors estimate is <span class="math inline">\(V\)</span>. Note
that it is always best to run GFA on all variants or an LD-pruned subset
and not on the most significant.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">maxz</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/apply.html" class="external-link">apply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">abs</a></span><span class="op">(</span><span class="va">Z_hat</span><span class="op">)</span>, <span class="fl">1</span>, <span class="va">max</span><span class="op">)</span></span>
<span><span class="va">Z_hat_top</span> <span class="op">&lt;-</span> <span class="va">Z_hat</span><span class="op">[</span><span class="va">maxz</span> <span class="op">&gt;</span> <span class="fl">5.45</span>,<span class="op">]</span> <span class="co">## corresponds to p &lt; 5e-8</span></span>
<span></span>
<span><span class="va">fit_svd</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/svd.html" class="external-link">svd</a></span><span class="op">(</span><span class="va">Z_hat_top</span><span class="op">)</span></span>
<span></span>
<span><span class="va">fit_svd</span><span class="op">$</span><span class="va">v</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../reference/plot_factors.html">plot_factors</a></span><span class="op">(</span><span class="va">.</span>, col_names <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"EF1"</span>, <span class="st">"EF2"</span>, <span class="st">"EF3"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_gradient.html" class="external-link">scale_fill_gradient2</a></span><span class="op">(</span>low <span class="op">=</span> <span class="fu"><a href="https://sjmgarnier.github.io/viridisLite/reference/viridis.html" class="external-link">viridis</a></span><span class="op">(</span><span class="fl">3</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, high <span class="op">=</span> <span class="fu"><a href="https://sjmgarnier.github.io/viridisLite/reference/viridis.html" class="external-link">viridis</a></span><span class="op">(</span><span class="fl">3</span><span class="op">)</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span>, name <span class="op">=</span> <span class="st">"Normalized\nEffect"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">xlab</a></span><span class="op">(</span><span class="st">"Mediator"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ylab</a></span><span class="op">(</span><span class="st">"Trait"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"Estimated Mediator Network (SVD)"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_discrete.html" class="external-link">scale_x_discrete</a></span><span class="op">(</span>position <span class="op">=</span> <span class="st">"top"</span><span class="op">)</span><span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>axis.text.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">0</span><span class="op">)</span>,</span>
<span>           panel.background <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_rect</a></span><span class="op">(</span>fill <span class="op">=</span> <span class="st">"white"</span><span class="op">)</span>,</span>
<span>           axis.ticks <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; Scale for <span style="color: #00BB00;">fill</span> is already present.</span></span>
<span><span class="co">#&gt; Adding another scale for <span style="color: #00BB00;">fill</span>, which will replace the existing scale.</span></span></code></pre></div>
<p><img src="gfa_with_simulated_data_files/figure-html/unnamed-chunk-8-1.png" width="700"></p>
<p>The first SVD factor is a good estimate of true factor F2 but the
subsequent columns do not correspond to true factors because they are
constrained to be orthogonal to the first.</p>
</div>
</div>
<div class="section level2">
<h2 id="more-complex-example">More Complex Example<a class="anchor" aria-label="anchor" href="#more-complex-example"></a>
</h2>
<p>Next we will generate and analyze some more complicated data
including LD and sample overlap. We use the LD pattern built into the
<code>GWASBrewer</code> package which is estimated from chromosome 19 of
the European subset of 1k Genomes.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"AF"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"ld_mat_list"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">ld_mat_list</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 39</span></span></code></pre></div>
<p>First we set up the simulation parameters. We will have 20 traits,
100k variants, and 5 unobserved factors. The GWAS sample size for the 20
traits will range between 30k and 50k with 30k samples common to GWAS of
all traits. The proportion of non-zero effects are set so that, on
average, there are 1k SNPs affecting each factor and 1k SNPs affecting
each trait directly.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">201</span><span class="op">)</span></span>
<span><span class="va">ntrait</span> <span class="op">&lt;-</span> <span class="fl">20</span></span>
<span><span class="va">nvar</span> <span class="op">&lt;-</span> <span class="fl">100000</span></span>
<span><span class="va">nfactor</span> <span class="op">&lt;-</span> <span class="fl">5</span></span>
<span></span>
<span><span class="co"># specify sample size matrix</span></span>
<span><span class="va">Nmat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fl">30000</span>, nrow <span class="op">=</span> <span class="va">ntrait</span>, ncol <span class="op">=</span> <span class="va">ntrait</span><span class="op">)</span></span>
<span><span class="va">Nunique</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">ceiling</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/Uniform.html" class="external-link">runif</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fl">20</span>, min <span class="op">=</span> <span class="fl">0</span>, max <span class="op">=</span> <span class="fl">20000</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/diag.html" class="external-link">diag</a></span><span class="op">(</span><span class="va">Nmat</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/diag.html" class="external-link">diag</a></span><span class="op">(</span><span class="va">Nmat</span><span class="op">)</span> <span class="op">+</span> <span class="va">Nunique</span></span>
<span></span>
<span><span class="co"># proportion of effect variants</span></span>
<span><span class="va">pi_L</span> <span class="op">&lt;-</span> <span class="va">pi_theta</span> <span class="op">&lt;-</span> <span class="fl">1000</span><span class="op">/</span><span class="va">nvar</span></span></code></pre></div>
<p>Next we generate a factor structure. To do this we use the
<code>generate_random_F</code> function in <code>GWASBrewer</code>. This
function requires a sampling function for drawing factor-trait effects,
the number of traits affected by each factor, the total trait
heritability, the proportion of heritability of each trait explained by
factors, and the heritability of the factors. For this example, we will
generate factor structures until we find one where every trait is
affected by at least one factor.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># sampling function</span></span>
<span><span class="va">g_F</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">n</span><span class="op">)</span><span class="op">{</span><span class="fu"><a href="https://rdrr.io/r/stats/Uniform.html" class="external-link">runif</a></span><span class="op">(</span><span class="va">n</span>, <span class="op">-</span><span class="fl">1</span>, <span class="fl">1</span><span class="op">)</span><span class="op">}</span></span>
<span><span class="co"># number of traits affected by each factor</span></span>
<span><span class="va">nz_factor</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">pmin</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/Poisson.html" class="external-link">rpois</a></span><span class="op">(</span><span class="va">nfactor</span>, <span class="fl">5</span><span class="op">)</span><span class="op">+</span><span class="fl">1</span>, <span class="va">ntrait</span><span class="op">)</span></span>
<span><span class="co"># heritability of each trait</span></span>
<span><span class="va">h2_trait</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/Uniform.html" class="external-link">runif</a></span><span class="op">(</span>n<span class="op">=</span><span class="va">ntrait</span>, <span class="fl">0.05</span>, <span class="fl">0.3</span><span class="op">)</span></span>
<span><span class="co"># proportion of trait variance explained by factors</span></span>
<span><span class="va">omega</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/Uniform.html" class="external-link">runif</a></span><span class="op">(</span>n<span class="op">=</span><span class="va">ntrait</span>, <span class="fl">0.5</span>, <span class="fl">1</span><span class="op">)</span></span>
<span><span class="co"># heritability of each factor</span></span>
<span><span class="va">h2_factor</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/Uniform.html" class="external-link">runif</a></span><span class="op">(</span>n<span class="op">=</span><span class="va">nfactor</span>, <span class="fl">0.5</span>, <span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">done</span> <span class="op">&lt;-</span> <span class="cn">FALSE</span></span>
<span><span class="kw">while</span><span class="op">(</span><span class="op">!</span><span class="va">done</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="co">#cat("going..")</span></span>
<span>  <span class="va">myF</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GWASBrewer/man/generate_random_F.html" class="external-link">generate_random_F</a></span><span class="op">(</span>K <span class="op">=</span> <span class="va">nfactor</span>, M <span class="op">=</span> <span class="va">ntrait</span>, g_F <span class="op">=</span> <span class="va">g_F</span>, </span>
<span>                         nz_factor <span class="op">=</span> <span class="va">nz_factor</span>, omega <span class="op">=</span> <span class="va">omega</span>, </span>
<span>                         h2_trait <span class="op">=</span> <span class="va">h2_trait</span><span class="op">)</span></span>
<span>  <span class="va">x</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">rowSums</a></span><span class="op">(</span><span class="va">myF</span><span class="op">^</span><span class="fl">2</span><span class="op">)</span></span>
<span>  <span class="kw">if</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/all.html" class="external-link">all</a></span><span class="op">(</span><span class="va">x</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span><span class="op">)</span> <span class="va">done</span> <span class="op">&lt;-</span> <span class="cn">TRUE</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p>Since GWAS samples overlap, the environmental correlation of the
traits will affect the final GWAS estimates. Since each of our factors
has some environmental component, some of the trait environmental
correlation is mediated by factors. We can specify the correlation of
the environmental components of the traits is not mediated by the
factors. For our example, we will use a block correlation structure.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">Rblock</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fl">0.5</span>, <span class="fl">4</span>, <span class="fl">4</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/diag.html" class="external-link">diag</a></span><span class="op">(</span><span class="va">Rblock</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fl">1</span></span>
<span><span class="va">R_E</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/kronecker.html" class="external-link">kronecker</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/diag.html" class="external-link">diag</a></span><span class="op">(</span><span class="fl">5</span><span class="op">)</span>, <span class="va">Rblock</span><span class="op">)</span></span></code></pre></div>
<p>Now we are ready to generate data.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dat2</span> <span class="op">&lt;-</span>  <span class="fu"><a href="https://rdrr.io/pkg/GWASBrewer/man/sim_lf.html" class="external-link">sim_lf</a></span><span class="op">(</span>F_mat <span class="op">=</span> <span class="va">myF</span>, N<span class="op">=</span><span class="va">Nmat</span>, J<span class="op">=</span><span class="va">nvar</span>, h2_trait <span class="op">=</span> <span class="va">h2_trait</span>,</span>
<span>            omega <span class="op">=</span> <span class="va">omega</span>, h2_factor <span class="op">=</span> <span class="va">h2_factor</span>,</span>
<span>            pi_L <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="va">pi_L</span>, <span class="va">nfactor</span><span class="op">)</span>, pi_theta <span class="op">=</span> <span class="va">pi_theta</span>,</span>
<span>            R_E <span class="op">=</span> <span class="va">R_E</span>, </span>
<span>            R_LD <span class="op">=</span> <span class="va">ld_mat_list</span>, af <span class="op">=</span> <span class="va">AF</span>, </span>
<span>            est_s <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> </span>
<span><span class="co">#&gt; SNP effects provided for 100000 SNPs and 20 traits.</span></span>
<span><span class="va">dat2</span> <span class="op">&lt;-</span> <span class="fu">GWASBrewer</span><span class="fu">:::</span><span class="fu">calc_ld_scores</span><span class="op">(</span><span class="va">dat2</span>, <span class="va">ld_mat_list</span><span class="op">)</span></span></code></pre></div>
<p>We can take a look at the generated true factor structure.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p</span> <span class="op">&lt;-</span> <span class="va">dat2</span><span class="op">$</span><span class="va">F_mat</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu">GFA</span><span class="fu">:::</span><span class="fu">norm_cols</span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/with.html" class="external-link">with</a></span><span class="op">(</span><span class="va">.</span>, <span class="va">A</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../reference/plot_factors.html">plot_factors</a></span><span class="op">(</span><span class="va">.</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_gradient.html" class="external-link">scale_fill_gradient2</a></span><span class="op">(</span>low <span class="op">=</span> <span class="fu"><a href="https://sjmgarnier.github.io/viridisLite/reference/viridis.html" class="external-link">viridis</a></span><span class="op">(</span><span class="fl">3</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, high <span class="op">=</span> <span class="fu"><a href="https://sjmgarnier.github.io/viridisLite/reference/viridis.html" class="external-link">viridis</a></span><span class="op">(</span><span class="fl">3</span><span class="op">)</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span>, name <span class="op">=</span> <span class="st">"Normalized\nEffect"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">xlab</a></span><span class="op">(</span><span class="st">"Mediator"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ylab</a></span><span class="op">(</span><span class="st">"Trait"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="co">#ggtitle("Mediator Network") +</span></span>
<span>  <span class="co">#scale_x_discrete(position = "top")+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>axis.text.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">0</span><span class="op">)</span>,</span>
<span>           panel.background <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_rect</a></span><span class="op">(</span>fill <span class="op">=</span> <span class="st">"white"</span><span class="op">)</span>,</span>
<span>           axis.ticks <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span>, </span>
<span>    axis.text <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">12</span><span class="op">)</span>,</span>
<span>    title <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">20</span><span class="op">)</span>,</span>
<span>        legend.position <span class="op">=</span> <span class="st">"none"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Scale for <span style="color: #00BB00;">fill</span> is already present.</span></span>
<span><span class="co">#&gt; Adding another scale for <span style="color: #00BB00;">fill</span>, which will replace the existing scale.</span></span>
<span><span class="va">p</span></span></code></pre></div>
<p><img src="gfa_with_simulated_data_files/figure-html/unnamed-chunk-14-1.png" width="700"></p>
<p>The simulated data also contains some objects describing the trait
covariance. The traits are scaled to have a total variance of 1.
<code>dat$Sigma_G</code> gives the genetic variance-covariance matrix,
so the diagonal of this matrix gives the heritability.
<code>dat$Sigma_E</code> gives the environmental variance-covariance
matrix, so the total trait correlation is
<code>dat$Sigma_E + dat$Sigma_G</code>. Useful for our purposes,
<code>dat$R</code> gives the residual correlation of the effect
estimates which is equal to <code>dat$trait_cor</code> with elements
scaled by the proportion of overlap between GWAS.</p>
<div class="section level3">
<h3 id="estimate-nuisance-correlation">Estimate Nuisance Correlation<a class="anchor" aria-label="anchor" href="#estimate-nuisance-correlation"></a>
</h3>
<p>Since we have sample overlap, we should estimate the residual
correlation. We can do this using either <code>R_ldsc</code> or
<code>R_pt</code> to use the LD-score regression or p-value thresholding
methods. For the LD-score regression method, we need to provide LD
scores which were computed by <code>GWASBrewer</code>.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">Z_hat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/with.html" class="external-link">with</a></span><span class="op">(</span><span class="va">dat2</span>, <span class="va">beta_hat</span><span class="op">/</span><span class="va">s_estimate</span><span class="op">)</span></span>
<span><span class="va">ldscores</span> <span class="op">&lt;-</span> <span class="va">dat2</span><span class="op">$</span><span class="va">snp_info</span><span class="op">$</span><span class="va">l2</span></span>
<span><span class="va">R1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/R_ldsc.html">R_ldsc</a></span><span class="op">(</span><span class="va">Z_hat</span>, ldscores <span class="op">=</span> <span class="va">ldscores</span>, ld_size <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">Z_hat</span><span class="op">)</span>, N <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/diag.html" class="external-link">diag</a></span><span class="op">(</span><span class="va">Nmat</span><span class="op">)</span><span class="op">)</span> </span>
<span><span class="va">R2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/with.html" class="external-link">with</a></span><span class="op">(</span><span class="va">dat2</span>, <span class="fu"><a href="../reference/R_pt.html">R_pt</a></span><span class="op">(</span><span class="va">beta_hat</span>, <span class="va">s_estimate</span>, p_val_thresh <span class="op">=</span> <span class="fl">0.05</span><span class="op">)</span><span class="op">)</span> </span></code></pre></div>
<p>We can take a look at the estimated <span class="math inline">\(R\)</span> vs the true <span class="math inline">\(R\)</span>. The LD-score method gives a more
accurate estimate but takes longer to compute.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_factors.html">plot_factors</a></span><span class="op">(</span><span class="va">dat2</span><span class="op">$</span><span class="va">R</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>axis.title <span class="op">=</span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"True R"</span><span class="op">)</span></span></code></pre></div>
<p><img src="gfa_with_simulated_data_files/figure-html/unnamed-chunk-16-1.png" width="700"></p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_factors.html">plot_factors</a></span><span class="op">(</span><span class="va">R1</span><span class="op">$</span><span class="va">Se</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>axis.title <span class="op">=</span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"Estimated R - LDSC"</span><span class="op">)</span></span></code></pre></div>
<p><img src="gfa_with_simulated_data_files/figure-html/unnamed-chunk-16-2.png" width="700"></p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_factors.html">plot_factors</a></span><span class="op">(</span><span class="va">R2</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>axis.title <span class="op">=</span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"Estimated R - p threshold"</span><span class="op">)</span></span></code></pre></div>
<p><img src="gfa_with_simulated_data_files/figure-html/unnamed-chunk-16-3.png" width="700"></p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">dat2</span><span class="op">$</span><span class="va">R</span>, <span class="va">R2</span>, xlab <span class="op">=</span> <span class="st">"True Correlation"</span>, ylab <span class="op">=</span> <span class="st">"Estimated Correlation - p threshold"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/abline.html" class="external-link">abline</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1</span><span class="op">)</span></span></code></pre></div>
<p><img src="gfa_with_simulated_data_files/figure-html/unnamed-chunk-16-4.png" width="700"></p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">dat2</span><span class="op">$</span><span class="va">R</span>, <span class="va">R1</span><span class="op">$</span><span class="va">Se</span>, xlab <span class="op">=</span> <span class="st">"True Correlation"</span>, ylab <span class="op">=</span> <span class="st">"Estimated Correlation - LDSC"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/abline.html" class="external-link">abline</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1</span><span class="op">)</span></span></code></pre></div>
<p><img src="gfa_with_simulated_data_files/figure-html/unnamed-chunk-16-5.png" width="700"></p>
<p>The LDSC estimate is closer to the truth, but GFA is not very
sensitive to this difference.</p>
</div>
<div class="section level3">
<h3 id="fit-gfa">Fit GFA<a class="anchor" aria-label="anchor" href="#fit-gfa"></a>
</h3>
<p>To fit GFA, we need to obtain an independent (LD-pruned) set of
variants. We can use <code>GWASBrewer</code> to find this for the
simulated data set. Note that we <strong>should not</strong> threshold
on p-value. It is ok to prioritize variants by minimum p-value for LD
pruning, but here we just prune randomly.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ix</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GWASBrewer/man/sim_ld_prune.html" class="external-link">sim_ld_prune</a></span><span class="op">(</span><span class="va">dat2</span>, R_LD <span class="op">=</span> <span class="va">ld_mat_list</span>, r2_thresh <span class="op">=</span> <span class="fl">0.01</span><span class="op">)</span></span>
<span><span class="co">#&gt; pvalue omitted so variants will be prioritized randomly</span></span>
<span><span class="va">Z_hat_indep</span> <span class="op">&lt;-</span> <span class="va">Z_hat</span><span class="op">[</span><span class="va">ix</span>,<span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">Z_hat_indep</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 4939   20</span></span></code></pre></div>
<p>Finally, we can fit GFA with or without adjustment for sample overlap
provided by the <code>R</code> argument. Recall that <code>R1</code> was
estimated using LD-score regression and <code>R2</code> was estimated
using p-value thresholding.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fit0</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/gfa_fit.html">gfa_fit</a></span><span class="op">(</span>Z_hat  <span class="op">=</span> <span class="va">Z_hat_indep</span>, N <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/diag.html" class="external-link">diag</a></span><span class="op">(</span><span class="va">Nmat</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; R is not supplied, fitting assuming full independence</span></span>
<span><span class="co">#&gt; Adding factor 1 to flash object...</span></span>
<span><span class="co">#&gt; Adding factor 2 to flash object...</span></span>
<span><span class="co">#&gt; Adding factor 3 to flash object...</span></span>
<span><span class="co">#&gt; Adding factor 4 to flash object...</span></span>
<span><span class="co">#&gt; Adding factor 5 to flash object...</span></span>
<span><span class="co">#&gt; Adding factor 6 to flash object...</span></span>
<span><span class="co">#&gt; Adding factor 7 to flash object...</span></span>
<span><span class="co">#&gt; Adding factor 8 to flash object...</span></span>
<span><span class="co">#&gt; Adding factor 9 to flash object...</span></span>
<span><span class="co">#&gt; Adding factor 10 to flash object...</span></span>
<span><span class="co">#&gt; Adding factor 11 to flash object...</span></span>
<span><span class="co">#&gt; Factor doesn't significantly increase objective and won't be added.</span></span>
<span><span class="co">#&gt; Wrapping up...</span></span>
<span><span class="co">#&gt; Done.</span></span>
<span><span class="co">#&gt; Backfitting 10 factors (tolerance: 1.47e-03)...</span></span>
<span><span class="co">#&gt;   Difference between iterations is within 1.0e+01...</span></span>
<span><span class="co">#&gt;   Difference between iterations is within 1.0e+00...</span></span>
<span><span class="co">#&gt;   Difference between iterations is within 1.0e-01...</span></span>
<span><span class="co">#&gt;   Difference between iterations is within 1.0e-02...</span></span>
<span><span class="co">#&gt;   Difference between iterations is within 1.0e-03...</span></span>
<span><span class="co">#&gt; Wrapping up...</span></span>
<span><span class="co">#&gt; Done.</span></span>
<span><span class="co">#&gt; Nullchecking 10 factors...</span></span>
<span><span class="co">#&gt; Done.</span></span>
<span><span class="co">#&gt; begin while</span></span>
<span><span class="co">#&gt; Nullchecking 10 factors...</span></span>
<span><span class="co">#&gt; Done.</span></span>
<span><span class="va">fit1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/gfa_fit.html">gfa_fit</a></span><span class="op">(</span>Z_hat <span class="op">=</span> <span class="va">Z_hat_indep</span>, N <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/diag.html" class="external-link">diag</a></span><span class="op">(</span><span class="va">Nmat</span><span class="op">)</span>,  R <span class="op">=</span> <span class="va">R1</span><span class="op">$</span><span class="va">Se</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning in check_R(R, dat$p, params): R has values different from 1 on the</span></span>
<span><span class="co">#&gt; diagonal. This is ok.</span></span>
<span><span class="co">#&gt; Adding factor 1 to flash object...</span></span>
<span><span class="co">#&gt; Adding factor 2 to flash object...</span></span>
<span><span class="co">#&gt; Adding factor 3 to flash object...</span></span>
<span><span class="co">#&gt; Adding factor 4 to flash object...</span></span>
<span><span class="co">#&gt; Adding factor 5 to flash object...</span></span>
<span><span class="co">#&gt; Adding factor 6 to flash object...</span></span>
<span><span class="co">#&gt; Adding factor 7 to flash object...</span></span>
<span><span class="co">#&gt; Adding factor 8 to flash object...</span></span>
<span><span class="co">#&gt; Adding factor 9 to flash object...</span></span>
<span><span class="co">#&gt; Adding factor 10 to flash object...</span></span>
<span><span class="co">#&gt; Adding factor 11 to flash object...</span></span>
<span><span class="co">#&gt; Adding factor 12 to flash object...</span></span>
<span><span class="co">#&gt; Factor doesn't significantly increase objective and won't be added.</span></span>
<span><span class="co">#&gt; Wrapping up...</span></span>
<span><span class="co">#&gt; Done.</span></span>
<span><span class="co">#&gt; Backfitting 28 factors (tolerance: 1.47e-03)...</span></span>
<span><span class="co">#&gt;   --Estimate of factor 5 is numerically zero!</span></span>
<span><span class="co">#&gt;   --Estimate of factor 6 is numerically zero!</span></span>
<span><span class="co">#&gt;   --Estimate of factor 7 is numerically zero!</span></span>
<span><span class="co">#&gt;   --Estimate of factor 8 is numerically zero!</span></span>
<span><span class="co">#&gt;   --Estimate of factor 10 is numerically zero!</span></span>
<span><span class="co">#&gt;   Difference between iterations is within 1.0e+04...</span></span>
<span><span class="co">#&gt;   Difference between iterations is within 1.0e+03...</span></span>
<span><span class="co">#&gt;   Difference between iterations is within 1.0e+02...</span></span>
<span><span class="co">#&gt;   Difference between iterations is within 1.0e+01...</span></span>
<span><span class="co">#&gt;   Difference between iterations is within 1.0e+00...</span></span>
<span><span class="co">#&gt;   Difference between iterations is within 1.0e-01...</span></span>
<span><span class="co">#&gt;   Difference between iterations is within 1.0e-02...</span></span>
<span><span class="co">#&gt;   Difference between iterations is within 1.0e-03...</span></span>
<span><span class="co">#&gt; Wrapping up...</span></span>
<span><span class="co">#&gt; Done.</span></span>
<span><span class="co">#&gt; Nullchecking 28 factors...</span></span>
<span><span class="co">#&gt;   5 factors are identically zero.</span></span>
<span><span class="co">#&gt; Factor22set to zero, increasing objective by 9.648e+00.</span></span>
<span><span class="co">#&gt; Factor26set to zero, increasing objective by 1.057e+01.</span></span>
<span><span class="co">#&gt; Factor27set to zero, increasing objective by 7.763e+00.</span></span>
<span><span class="co">#&gt; Factor28set to zero, increasing objective by 1.118e+01.</span></span>
<span><span class="co">#&gt; Wrapping up...</span></span>
<span><span class="co">#&gt;   Removed 9 factors.</span></span>
<span><span class="co">#&gt; Done.</span></span>
<span><span class="co">#&gt; begin while</span></span>
<span><span class="co">#&gt; Nullchecking 19 factors...</span></span>
<span><span class="co">#&gt; Factor6set to zero, increasing objective by 1.623e+02.</span></span>
<span><span class="co">#&gt; Factor18set to zero, increasing objective by 4.582e+00.</span></span>
<span><span class="co">#&gt; Factor19set to zero, increasing objective by 6.542e+00.</span></span>
<span><span class="co">#&gt; Wrapping up...</span></span>
<span><span class="co">#&gt;   Removed 3 factors.</span></span>
<span><span class="co">#&gt; Done.</span></span>
<span><span class="va">fit2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/gfa_fit.html">gfa_fit</a></span><span class="op">(</span>Z_hat <span class="op">=</span> <span class="va">Z_hat_indep</span>, N <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/diag.html" class="external-link">diag</a></span><span class="op">(</span><span class="va">Nmat</span><span class="op">)</span>,  R <span class="op">=</span> <span class="va">R2</span><span class="op">)</span></span>
<span><span class="co">#&gt; Adding factor 1 to flash object...</span></span>
<span><span class="co">#&gt; Adding factor 2 to flash object...</span></span>
<span><span class="co">#&gt; Adding factor 3 to flash object...</span></span>
<span><span class="co">#&gt; Adding factor 4 to flash object...</span></span>
<span><span class="co">#&gt; Adding factor 5 to flash object...</span></span>
<span><span class="co">#&gt; Adding factor 6 to flash object...</span></span>
<span><span class="co">#&gt; Adding factor 7 to flash object...</span></span>
<span><span class="co">#&gt; Adding factor 8 to flash object...</span></span>
<span><span class="co">#&gt; Adding factor 9 to flash object...</span></span>
<span><span class="co">#&gt; Adding factor 10 to flash object...</span></span>
<span><span class="co">#&gt; Adding factor 11 to flash object...</span></span>
<span><span class="co">#&gt; Adding factor 12 to flash object...</span></span>
<span><span class="co">#&gt; Factor doesn't significantly increase objective and won't be added.</span></span>
<span><span class="co">#&gt; Wrapping up...</span></span>
<span><span class="co">#&gt; Done.</span></span>
<span><span class="co">#&gt; Backfitting 28 factors (tolerance: 1.47e-03)...</span></span>
<span><span class="co">#&gt;   --Estimate of factor 8 is numerically zero!</span></span>
<span><span class="co">#&gt;   --Estimate of factor 10 is numerically zero!</span></span>
<span><span class="co">#&gt;   --Estimate of factor 5 is numerically zero!</span></span>
<span><span class="co">#&gt;   --Estimate of factor 6 is numerically zero!</span></span>
<span><span class="co">#&gt;   --Estimate of factor 7 is numerically zero!</span></span>
<span><span class="co">#&gt;   Difference between iterations is within 1.0e+04...</span></span>
<span><span class="co">#&gt;   Difference between iterations is within 1.0e+03...</span></span>
<span><span class="co">#&gt;   Difference between iterations is within 1.0e+02...</span></span>
<span><span class="co">#&gt;   Difference between iterations is within 1.0e+01...</span></span>
<span><span class="co">#&gt;   Difference between iterations is within 1.0e+00...</span></span>
<span><span class="co">#&gt;   Difference between iterations is within 1.0e-01...</span></span>
<span><span class="co">#&gt;   Difference between iterations is within 1.0e-02...</span></span>
<span><span class="co">#&gt;   Difference between iterations is within 1.0e-03...</span></span>
<span><span class="co">#&gt; Wrapping up...</span></span>
<span><span class="co">#&gt; Done.</span></span>
<span><span class="co">#&gt; Nullchecking 28 factors...</span></span>
<span><span class="co">#&gt;   5 factors are identically zero.</span></span>
<span><span class="co">#&gt; Factor22set to zero, increasing objective by 3.888e+01.</span></span>
<span><span class="co">#&gt; Factor23set to zero, increasing objective by 3.826e+01.</span></span>
<span><span class="co">#&gt; Factor24set to zero, increasing objective by 3.264e+01.</span></span>
<span><span class="co">#&gt; Factor25set to zero, increasing objective by 4.091e+01.</span></span>
<span><span class="co">#&gt; Factor26set to zero, increasing objective by 2.425e+01.</span></span>
<span><span class="co">#&gt; Factor27set to zero, increasing objective by 3.229e+01.</span></span>
<span><span class="co">#&gt; Factor28set to zero, increasing objective by 2.313e+01.</span></span>
<span><span class="co">#&gt; Wrapping up...</span></span>
<span><span class="co">#&gt;   Removed 12 factors.</span></span>
<span><span class="co">#&gt; Done.</span></span>
<span><span class="co">#&gt; begin while</span></span>
<span><span class="co">#&gt; Nullchecking 16 factors...</span></span>
<span><span class="co">#&gt; Factor6set to zero, increasing objective by 1.067e+02.</span></span>
<span><span class="co">#&gt; Wrapping up...</span></span>
<span><span class="co">#&gt;   Removed one factor.</span></span>
<span><span class="co">#&gt; Done.</span></span></code></pre></div>
<p>We can compare these results to fitting GFA using the true nuisance
correlation matrix,</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fitoracle</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/gfa_fit.html">gfa_fit</a></span><span class="op">(</span>Z_hat  <span class="op">=</span> <span class="va">Z_hat_indep</span>, N <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/diag.html" class="external-link">diag</a></span><span class="op">(</span><span class="va">Nmat</span><span class="op">)</span>, R <span class="op">=</span> <span class="va">dat2</span><span class="op">$</span><span class="va">R</span><span class="op">)</span></span>
<span><span class="co">#&gt; Adding factor 1 to flash object...</span></span>
<span><span class="co">#&gt; Adding factor 2 to flash object...</span></span>
<span><span class="co">#&gt; Adding factor 3 to flash object...</span></span>
<span><span class="co">#&gt; Adding factor 4 to flash object...</span></span>
<span><span class="co">#&gt; Adding factor 5 to flash object...</span></span>
<span><span class="co">#&gt; Adding factor 6 to flash object...</span></span>
<span><span class="co">#&gt; Adding factor 7 to flash object...</span></span>
<span><span class="co">#&gt; Adding factor 8 to flash object...</span></span>
<span><span class="co">#&gt; Adding factor 9 to flash object...</span></span>
<span><span class="co">#&gt; Adding factor 10 to flash object...</span></span>
<span><span class="co">#&gt; Adding factor 11 to flash object...</span></span>
<span><span class="co">#&gt; Adding factor 12 to flash object...</span></span>
<span><span class="co">#&gt; Factor doesn't significantly increase objective and won't be added.</span></span>
<span><span class="co">#&gt; Wrapping up...</span></span>
<span><span class="co">#&gt; Done.</span></span>
<span><span class="co">#&gt; Backfitting 28 factors (tolerance: 1.47e-03)...</span></span>
<span><span class="co">#&gt;   --Estimate of factor 5 is numerically zero!</span></span>
<span><span class="co">#&gt;   --Estimate of factor 6 is numerically zero!</span></span>
<span><span class="co">#&gt;   --Estimate of factor 7 is numerically zero!</span></span>
<span><span class="co">#&gt;   --Estimate of factor 8 is numerically zero!</span></span>
<span><span class="co">#&gt;   --Estimate of factor 10 is numerically zero!</span></span>
<span><span class="co">#&gt;   Difference between iterations is within 1.0e+04...</span></span>
<span><span class="co">#&gt;   Difference between iterations is within 1.0e+03...</span></span>
<span><span class="co">#&gt;   Difference between iterations is within 1.0e+02...</span></span>
<span><span class="co">#&gt;   Difference between iterations is within 1.0e+01...</span></span>
<span><span class="co">#&gt;   Difference between iterations is within 1.0e+00...</span></span>
<span><span class="co">#&gt;   Difference between iterations is within 1.0e-01...</span></span>
<span><span class="co">#&gt;   Difference between iterations is within 1.0e-02...</span></span>
<span><span class="co">#&gt; Wrapping up...</span></span>
<span><span class="co">#&gt; Done.</span></span>
<span><span class="co">#&gt; Nullchecking 28 factors...</span></span>
<span><span class="co">#&gt;   5 factors are identically zero.</span></span>
<span><span class="co">#&gt; Factor24set to zero, increasing objective by 6.277e+00.</span></span>
<span><span class="co">#&gt; Factor26set to zero, increasing objective by 1.314e+01.</span></span>
<span><span class="co">#&gt; Factor27set to zero, increasing objective by 9.652e+00.</span></span>
<span><span class="co">#&gt; Factor28set to zero, increasing objective by 2.052e+01.</span></span>
<span><span class="co">#&gt; Wrapping up...</span></span>
<span><span class="co">#&gt;   Removed 9 factors.</span></span>
<span><span class="co">#&gt; Done.</span></span>
<span><span class="co">#&gt; begin while</span></span>
<span><span class="co">#&gt; Nullchecking 19 factors...</span></span>
<span><span class="co">#&gt; Factor6set to zero, increasing objective by 1.578e+02.</span></span>
<span><span class="co">#&gt; Factor17set to zero, increasing objective by 6.327e+00.</span></span>
<span><span class="co">#&gt; Factor18set to zero, increasing objective by 3.836e+00.</span></span>
<span><span class="co">#&gt; Factor19set to zero, increasing objective by 1.115e+01.</span></span>
<span><span class="co">#&gt; Wrapping up...</span></span>
<span><span class="co">#&gt;   Removed 4 factors.</span></span>
<span><span class="co">#&gt; Done.</span></span></code></pre></div>
<p>In this example, we get nearly identical results using either
estimate of R or the truth. With any of these settings, we are able to
accurately recover all five factors. We can see this by using the
<code>min_norm</code> function which compares two factor matrices
finding the rotation, <span class="math inline">\(Q\)</span> that
minimized the frobenious norm <span class="math inline">\(\vert \vert
F_1 - Q F_2 \vert \vert_{F}\)</span>. The <code>min_norm</code> function
will report which factors match between the matrices, how well they
match, as well as the overall frobenious norm. A frobenious norm of 0
means the matrices are exactly the same. The value of the frobenious
norm can be roughly interpreted as the number of factors in <span class="math inline">\(F_2\)</span> with no match in <span class="math inline">\(F_1\)</span> plus the number of factors in <span class="math inline">\(F_1\)</span> with no match in <span class="math inline">\(F_2\)</span>, with partial scores for partially
matching factors.</p>
<p>For example, we can first compare the true factor structure to our
first GFA estimate. This shows that all five factors were recovered with
very high correlation. The value in the <code>val</code> column of the
<code>solution</code> element is equal to the normalized inner product
of the two factors, approximately equal to the correlation.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/min_norm.html">min_norm</a></span><span class="op">(</span>f_true <span class="op">=</span> <span class="va">dat2</span><span class="op">$</span><span class="va">F_mat</span>, f_hat <span class="op">=</span> <span class="va">fit1</span><span class="op">$</span><span class="va">F_hat</span><span class="op">)</span></span>
<span><span class="co">#&gt; $solution</span></span>
<span><span class="co">#&gt;   true_ix est_ix max_true_val max_hat_val     penalty match_score</span></span>
<span><span class="co">#&gt; 1       5      4    0.5387060   0.5291128 0.001436740   0.9992816</span></span>
<span><span class="co">#&gt; 2       1      2    0.5082579   0.4920455 0.003164712   0.9984176</span></span>
<span><span class="co">#&gt; 3       4      1    0.5321195   0.5697020 0.005522261   0.9972389</span></span>
<span><span class="co">#&gt; 4       3      3    0.5828887   0.5819004 0.013035870   0.9934821</span></span>
<span><span class="co">#&gt; 5       2      5    0.7047004   0.7606438 0.015395358   0.9923023</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $frob_n</span></span>
<span><span class="co">#&gt; [1] 0.1963541</span></span></code></pre></div>
<p>Looking at the frobenious norm values, we can see that all GFA
estimates that included the nuisance correlation were very similar.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/min_norm.html">min_norm</a></span><span class="op">(</span>f_true <span class="op">=</span> <span class="va">dat2</span><span class="op">$</span><span class="va">F_mat</span>, f_hat <span class="op">=</span> <span class="va">fit1</span><span class="op">$</span><span class="va">F_hat</span><span class="op">)</span><span class="op">$</span><span class="va">frob_n</span></span>
<span><span class="co">#&gt; [1] 0.1963541</span></span>
<span><span class="fu"><a href="../reference/min_norm.html">min_norm</a></span><span class="op">(</span>f_true <span class="op">=</span> <span class="va">dat2</span><span class="op">$</span><span class="va">F_mat</span>, f_hat <span class="op">=</span> <span class="va">fit2</span><span class="op">$</span><span class="va">F_hat</span><span class="op">)</span><span class="op">$</span><span class="va">frob_n</span></span>
<span><span class="co">#&gt; [1] 0.2314396</span></span>
<span><span class="fu"><a href="../reference/min_norm.html">min_norm</a></span><span class="op">(</span>f_true <span class="op">=</span> <span class="va">dat2</span><span class="op">$</span><span class="va">F_mat</span>, f_hat <span class="op">=</span> <span class="va">fitoracle</span><span class="op">$</span><span class="va">F_hat</span><span class="op">)</span><span class="op">$</span><span class="va">frob_n</span></span>
<span><span class="co">#&gt; [1] 0.1906022</span></span></code></pre></div>
<p>If we compare these results with the results where we omitted the
<code>R</code> argument, we can see that not adjusting for the nuisance
correlation has resulted in four extra factors being estimated that
donâ€™t match any true factors. The correlation between the matching
factors is also slightly lower than in the result where we did account
for sample overlap. The two factors given scores of NA are single trait
factors and donâ€™t count towards the overall forbenious norm score. The
value in <code>opt_frob_n</code> shows the best score achievable by
selecting only a subset of the estimated factors. We can see that, even
if we knew which factors to remove, we would still have a worse match
than achieved by the method which accounts for sample overlap.</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/min_norm.html">min_norm</a></span><span class="op">(</span>f_true <span class="op">=</span> <span class="va">dat2</span><span class="op">$</span><span class="va">F_mat</span>, f_hat <span class="op">=</span> <span class="va">fit0</span><span class="op">$</span><span class="va">F_hat</span><span class="op">)</span></span>
<span><span class="co">#&gt; There are  1  single trait factors in f_hat</span></span>
<span><span class="co">#&gt; $solution</span></span>
<span><span class="co">#&gt;    true_ix est_ix max_true_val max_hat_val     penalty match_score</span></span>
<span><span class="co">#&gt; 1        4      1    0.5321195   0.5486638 0.002814011   0.9985930</span></span>
<span><span class="co">#&gt; 2        5      4    0.5387060   0.5310407 0.004892544   0.9975537</span></span>
<span><span class="co">#&gt; 3        3      3    0.5828887   0.5858742 0.014126643   0.9929367</span></span>
<span><span class="co">#&gt; 4        1      2    0.5082579   0.5023813 0.070248731   0.9648756</span></span>
<span><span class="co">#&gt; 5        2      7    0.7047004   0.7111798 0.097523443   0.9512383</span></span>
<span><span class="co">#&gt; 6       NA      9           NA   0.4175495 1.000000000   0.0000000</span></span>
<span><span class="co">#&gt; 7       NA      8           NA   0.4232410 1.000000000   0.0000000</span></span>
<span><span class="co">#&gt; 8       NA      5           NA   0.3864060 1.000000000   0.0000000</span></span>
<span><span class="co">#&gt; 9       NA      6           NA   0.5245608 1.000000000   0.0000000</span></span>
<span><span class="co">#&gt; 10      NA     10           NA   0.9999966          NA          NA</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $frob_n</span></span>
<span><span class="co">#&gt; [1] 2.046853</span></span></code></pre></div>
<p>Plotting the factors estimated with and without the correction for
overlapping samples, shows that the factor structure of the extra
factors mimics the structure of the environmental correlation. We can
also see that some of the matching factors estimated in
<code>fit0</code> are contaminated with some of the structure of the
environmental correlation. Here we plot only the multi-trait
factors.</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p0</span> <span class="op">&lt;-</span> <span class="va">fit0</span><span class="op">$</span><span class="va">F_hat_multi</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="../reference/plot_factors.html">plot_factors</a></span><span class="op">(</span><span class="va">.</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_gradient.html" class="external-link">scale_fill_gradient2</a></span><span class="op">(</span>low <span class="op">=</span> <span class="fu"><a href="https://sjmgarnier.github.io/viridisLite/reference/viridis.html" class="external-link">viridis</a></span><span class="op">(</span><span class="fl">3</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, high <span class="op">=</span> <span class="fu"><a href="https://sjmgarnier.github.io/viridisLite/reference/viridis.html" class="external-link">viridis</a></span><span class="op">(</span><span class="fl">3</span><span class="op">)</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">xlab</a></span><span class="op">(</span><span class="st">"Factor"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ylab</a></span><span class="op">(</span><span class="st">"Trait"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"Without Sample Overlap Correction"</span><span class="op">)</span> </span>
<span><span class="co">#&gt; Scale for <span style="color: #00BB00;">fill</span> is already present.</span></span>
<span><span class="co">#&gt; Adding another scale for <span style="color: #00BB00;">fill</span>, which will replace the existing scale.</span></span>
<span></span>
<span><span class="va">p1</span> <span class="op">&lt;-</span> <span class="va">fit1</span><span class="op">$</span><span class="va">F_hat_multi</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="../reference/plot_factors.html">plot_factors</a></span><span class="op">(</span><span class="va">.</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_gradient.html" class="external-link">scale_fill_gradient2</a></span><span class="op">(</span>low <span class="op">=</span> <span class="fu"><a href="https://sjmgarnier.github.io/viridisLite/reference/viridis.html" class="external-link">viridis</a></span><span class="op">(</span><span class="fl">3</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, high <span class="op">=</span> <span class="fu"><a href="https://sjmgarnier.github.io/viridisLite/reference/viridis.html" class="external-link">viridis</a></span><span class="op">(</span><span class="fl">3</span><span class="op">)</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">xlab</a></span><span class="op">(</span><span class="st">"Factor"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ylab</a></span><span class="op">(</span><span class="st">"Trait"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"With Sample Overlap Correction"</span><span class="op">)</span> </span>
<span><span class="co">#&gt; Scale for <span style="color: #00BB00;">fill</span> is already present.</span></span>
<span><span class="co">#&gt; Adding another scale for <span style="color: #00BB00;">fill</span>, which will replace the existing scale.</span></span>
<span><span class="va">p0</span></span></code></pre></div>
<p><img src="gfa_with_simulated_data_files/figure-html/unnamed-chunk-23-1.png" width="700"></p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p1</span></span></code></pre></div>
<p><img src="gfa_with_simulated_data_files/figure-html/unnamed-chunk-23-2.png" width="700"></p>
</div>
</div>
<div class="section level2">
<h2 id="understanding-scaling-within-and-between-factors">Understanding Scaling Within and Between Factors<a class="anchor" aria-label="anchor" href="#understanding-scaling-within-and-between-factors"></a>
</h2>
<p>If data is supplied as z-scores and sample sizes, the effects in
<code>F_hat</code> will be returned are on the standardized effect
scale. For example, if the first column of <code>F_hat</code> was equal
to <span class="math inline">\((1, 2, 0)\)</span>, this would mean that,
for each unit increase in the first factor, there is a 1 sd increase in
trait 1, a 2 sd increase in trait 2 and no change in trait 3. However
the scale of the inferred factor is arbitrary, so a factor <span class="math inline">\((1, 2, 0)\)</span> is equivalent to a factor <span class="math inline">\((0.1, 0.2, 0)\)</span>. This means that an
appropriate interpretation of the factor <span class="math inline">\((1,
2, 0)\)</span> is that the first factor has an effect on trait 2 that is
twice as big as its effect on trait 1 <strong>in units of trait standard
deviations</strong>. For plotting convenience, in our output, we always
scale the factors so that the columns of <code>F_hat</code> have norm
1.</p>
<p>If only z-scores and no sample sizes are provided, the effects in
<code>F_hat</code> will be returned are on the z-score scale,
proportional to trait standard deviation divided times square root of
sample size. Since sample size is a study design parameter, these
effects are not immediately interpretable, but could be converted to the
standardized trait scale by dividing each row of <code>F_hat</code> by
the square root of the sample size for the corresponding trait.</p>
<p>If effect estimates and standard errors are supplied, the the effects
in <code>F_hat</code> will be returned on the natural scale, i.e.Â the
scale of the trait used in GWAS. Alternatively, if effects on the
natural scale are desired, GFA can be fit using z-scores, and then
effects can be transformed. Supplying data as z-scores and sample size
or as effect estimates and standard errors should give nearly identical
results up to trait scaling.</p>
</div>
<div class="section level2">
<h2 id="percent-of-heritability-explained">Percent of Heritability Explained<a class="anchor" aria-label="anchor" href="#percent-of-heritability-explained"></a>
</h2>
<p>The <code>gfa_pve</code> element of the GFA results object contains
two elements, a vector, <code>genet_var</code> and a matrix
<code>pve</code>. The <code>genet_var</code> vector gives the proportion
of trait variance explained by the variants that GFA was fit with
according to the fitted model. This is an estimate of the heritability
attributable to the variants in the model. In most cases, this is an
underestimate of the total trait heritability because GFA is only fit
with an LD-pruned subset of variants. We donâ€™t recommend GFA as an ideal
heritability estimation tool. The more interesting object is
<code>pve</code> which has dimension traits by inferred factors and
gives for each trait, the proportion of heritability explained by that
factor. We have found that even though the model does not capture all of
the total heritability, our estimates of proportion of heritability
explained by each factor are close to the truth.</p>
<p>Using results from the second example, we plot the proportion of
variance explained matrix:</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p1</span> <span class="op">&lt;-</span> <span class="va">fit1</span><span class="op">$</span><span class="va">gfa_pve</span><span class="op">$</span><span class="va">pve</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="../reference/plot_factors.html">plot_factors</a></span><span class="op">(</span><span class="va">.</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_gradient.html" class="external-link">scale_fill_gradientn</a></span><span class="op">(</span>colors <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"white"</span>, <span class="fu"><a href="https://rdrr.io/r/base/rev.html" class="external-link">rev</a></span><span class="op">(</span><span class="fu"><a href="https://sjmgarnier.github.io/viridisLite/reference/viridis.html" class="external-link">viridis</a></span><span class="op">(</span><span class="fl">3</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">xlab</a></span><span class="op">(</span><span class="st">"Factor"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ylab</a></span><span class="op">(</span><span class="st">"Trait"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Scale for <span style="color: #00BB00;">fill</span> is already present.</span></span>
<span><span class="co">#&gt; Adding another scale for <span style="color: #00BB00;">fill</span>, which will replace the existing scale.</span></span>
<span><span class="va">p1</span></span></code></pre></div>
<p><img src="gfa_with_simulated_data_files/figure-html/unnamed-chunk-24-1.png" width="700"></p>
<p>We can compare our estimated proportion of heritability explained
with the true heritability explained by each factor.</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sim_calc_pve</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">dat</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="va">sj</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">dat</span><span class="op">$</span><span class="va">F_mat</span><span class="op">)</span>, <span class="kw">function</span><span class="op">(</span><span class="va">kk</span><span class="op">)</span><span class="op">{</span></span>
<span>     <span class="fu"><a href="https://rdrr.io/pkg/GWASBrewer/man/compute_h2.html" class="external-link">compute_h2</a></span><span class="op">(</span>b_joint <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/with.html" class="external-link">with</a></span><span class="op">(</span><span class="va">dat</span>, <span class="va">L_mat_joint</span><span class="op">[</span>,<span class="va">kk</span>,drop <span class="op">=</span> <span class="cn">F</span><span class="op">]</span> <span class="op"><a href="https://rdrr.io/r/base/matmult.html" class="external-link">%*%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="va">F_mat</span><span class="op">[</span>,<span class="va">kk</span>, drop <span class="op">=</span> <span class="cn">F</span><span class="op">]</span><span class="op">)</span><span class="op">)</span>, </span>
<span>                geno_scale <span class="op">=</span> <span class="st">"allele"</span>, </span>
<span>                af <span class="op">=</span> <span class="va">AF</span>, </span>
<span>                R_LD <span class="op">=</span> <span class="va">ld_mat_list</span><span class="op">)</span></span>
<span>    <span class="op">}</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span>nrow <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">dat</span><span class="op">$</span><span class="va">F_mat</span><span class="op">)</span>, byrow <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/function.html" class="external-link">return</a></span><span class="op">(</span><span class="va">sj</span><span class="op">/</span><span class="va">dat</span><span class="op">$</span><span class="va">h2</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="va">pve_true</span> <span class="op">&lt;-</span> <span class="fu">sim_calc_pve</span><span class="op">(</span><span class="va">dat2</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## order true pve and estimated pve so they can be compared</span></span>
<span><span class="va">sol</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/min_norm.html">min_norm</a></span><span class="op">(</span>f_hat <span class="op">=</span> <span class="va">fit1</span><span class="op">$</span><span class="va">F_hat</span>, </span>
<span>                f_true <span class="op">=</span> <span class="va">dat2</span><span class="op">$</span><span class="va">F_mat</span><span class="op">)</span><span class="op">$</span><span class="va">solution</span></span>
<span><span class="va">order_true</span> <span class="op">&lt;-</span> <span class="va">sol</span><span class="op">$</span><span class="va">true_ix</span></span>
<span><span class="va">order_est</span> <span class="op">&lt;-</span> <span class="va">sol</span><span class="op">$</span><span class="va">est_ix</span></span>
<span><span class="va">pve_est</span> <span class="op">&lt;-</span> <span class="va">fit1</span><span class="op">$</span><span class="va">gfa_pve</span><span class="op">$</span><span class="va">pve</span><span class="op">[</span>, <span class="va">order_est</span><span class="op">]</span></span>
<span><span class="va">pve_true</span> <span class="op">&lt;-</span> <span class="va">pve_true</span><span class="op">[</span>, <span class="va">order_true</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">pve_true</span>, <span class="va">pve_est</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/abline.html" class="external-link">abline</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1</span><span class="op">)</span></span></code></pre></div>
<p><img src="gfa_with_simulated_data_files/figure-html/unnamed-chunk-25-1.png" width="700"></p>
<p>In this example, our estimated proportions of heritability explained
tend to be lower than the truth, but are generally fairly close.</p>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Jean Morrison.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
