<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="Cross-Trait LD score regression"><title>Cross-Trait LD score regression — ldsc_rg • GFA</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.11/clipboard.min.js" integrity="sha512-7O5pXpc0oCRrxk8RUfDYFgn0nO1t+jLuIOQdOMRp4APB7uZ4vSjspzp5y6YDtDs4VzUSTbWzBFZ/LKJhnyFOKw==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Cross-Trait LD score regression — ldsc_rg"><meta property="og:description" content="Cross-Trait LD score regression"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light" data-bs-theme="light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">GFA</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.2.0333</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/gfa_with_simulated_data.html">Running GFA with Simulated Data</a>
  </div>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"></ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Cross-Trait LD score regression</h1>
      
      <div class="d-none name"><code>ldsc_rg.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Cross-Trait LD score regression</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">ldsc_rg</span><span class="op">(</span></span>
<span>  <span class="va">ld_score</span>,</span>
<span>  <span class="va">ld_size</span>,</span>
<span>  <span class="va">z1</span>,</span>
<span>  <span class="va">z2</span>,</span>
<span>  <span class="va">sample_size_1</span>,</span>
<span>  <span class="va">sample_size_2</span>,</span>
<span>  blocks <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  h2_1 <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  h2_2 <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  intercept <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  intercept_h2_1 <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  intercept_h2_2 <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  step1_chisq_max <span class="op">=</span> <span class="fl">30</span>,</span>
<span>  chi2_thr2 <span class="op">=</span> <span class="cn">Inf</span>,</span>
<span>  ncores <span class="op">=</span> <span class="fl">1</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>ld_score</dt>
<dd><p>Vector of LD scores.</p></dd>


<dt>ld_size</dt>
<dd><p>Number of variants used to compute `ld_score`.</p></dd>


<dt>z1</dt>
<dd><p>Vector of z-scores for trait 1.</p></dd>


<dt>z2</dt>
<dd><p>Vector of z-scores for trait 2.</p></dd>


<dt>sample_size_1</dt>
<dd><p>Sample size of GWAS for trait 1.
Possibly a vector, or just a single value.</p></dd>


<dt>sample_size_2</dt>
<dd><p>Sample size of GWAS for trait 2.
Possibly a vector, or just a single value.</p></dd>


<dt>blocks</dt>
<dd><p>Either a single number specifying the number of blocks,
or a vector of integers specifying the block number of each `chi2` value.
Default is `200` for `snp_ldsc()`, dividing into 200 blocks of approximately
equal size. `NULL` can also be used to skip estimating standard errors,
which is the default for `snp_ldsc2()`.</p></dd>


<dt>intercept</dt>
<dd><p>You can constrain the intercept to some value (e.g. 0).
Default is `NULL` (the intercept is estimated).
Use a value of 0 if you are sure there is no overlap between GWAS samples.</p></dd>


<dt>intercept_h2_1</dt>
<dd><p>Intercept for heritability of trait 1 (default is NULL so the intercept is estimated).</p></dd>


<dt>intercept_h2_2</dt>
<dd><p>Intercept for heritability of trait 2 (default is NULL so the intercept is estimated).</p></dd>


<dt>step1_chisq_max</dt>
<dd><p>Threshold on `chi2` in step 1. Default is `30`.</p></dd>


<dt>chi2_thr2</dt>
<dd><p>Threshold on `chi2` in step 2. Default is `Inf` (none).</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    

<p>Vector of 4 values (or only the first 2 if `blocks = NULL`):
 - `[["int"]]`: LDSC regression intercept,
 - `[["int_se"]]`: SE of this intercept,
 - `[["h2"]]`: LDSC regression estimate of (SNP) heritability
 - `[["h2_se"]]`: SE of this heritability estimate.</p>
    </div>

  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Jean Morrison.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

    </footer></div>

  

  

  </body></html>

